### PR TITLE
Feat/handle drop shadow

### PR DESCRIPTION
**Pull request by mathieurobert1**
Handle drop shadow for point and directional lights:

image

### CODE REVIEW

File: include/Primitives/Cone.hpp [hunk 1]
```diff
@@ -105,6 +96,14 @@ namespace Primitive {
              */
             std::shared_ptr<Material::IMaterial> getMaterial() const;
 
+            /**
+             * @brief Get the Normal of the object
```
Comment (AxelF44) [inline include/Primitives/Cone.hpp:100]:
```suggestion
             * @brief Get the Normal of the object.
```

File: include/Primitives/Cylinder.hpp [hunk 2]
```diff
@@ -93,6 +84,15 @@ namespace Primitive {
              * @param material Material of cylinder
              */
             void setMaterial(std::shared_ptr<Material::IMaterial> material);
+
+            /**
+             * @brief Get the Normal of the object
```
Comment (AxelF44) [inline include/Primitives/Cylinder.hpp:89]:
```suggestion
             * @brief Get the Normal of the object.
```

File: include/Primitives/IPrimitive.hpp [hunk 3]
```diff
@@ -40,12 +42,18 @@ namespace Primitive {
             virtual Math::Point3D hitPoint(const Raytracer::Ray& ray) const = 0;
 
             /**
-             * @brief compute the hit point color of a primitive
+             * @brief Get the Normal of the object
```
Comment (AxelF44) [inline include/Primitives/IPrimitive.hpp:45]:
```suggestion
             * @brief Get the Normal of the object.
```

File: include/Primitives/Plane.hpp [hunk 4]
```diff
@@ -78,7 +78,7 @@ namespace Primitive {
              * @brief Get the Material object
```
Comment (AxelF44) [inline include/Primitives/Plane.hpp:78]:
```suggestion
             * @brief Get the Material object.
```

File: include/Primitives/Plane.hpp [hunk 5]
```diff
@@ -78,7 +78,7 @@ namespace Primitive {
...
              * @return Material of plane
              */
-            std::shared_ptr<Material::IMaterial> getMaterial() const;
+            std::shared_ptr<Material::IMaterial> getMaterial() const override;
 
             /**
              * @brief Set the Material
```
Comment (AxelF44) [inline include/Primitives/Plane.hpp:84]:
```suggestion
             * @brief Set the Material.
```

File: include/Primitives/PrimitivesContainer.hpp [hunk 6]
```diff
@@ -59,6 +60,16 @@ namespace Primitive {
              */
             std::vector<std::shared_ptr<Primitive::IPrimitive>> getPrimitivesList(void) const;
 
+            /**
+             * @brief compute the color pixel of a primitive's hitpoint
```
Comment (AxelF44) [inline include/Primitives/PrimitivesContainer.hpp:64]:
```suggestion
             * @brief Compute the color pixel of a primitive's hitpoint.
```

File: include/Primitives/Shadow.hpp [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+namespace Primitives {
+    class Shadow;
+};
+
+class Primitives::Shadow {
+
+    public:
+
+        /**
+         * @brief Construct a new Shadow object
```
Comment (AxelF44) [inline include/Primitives/Shadow.hpp:24]:
```suggestion
         * @brief Construct a new Shadow object.
```

File: include/Primitives/Shadow.hpp [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+         *
+         * @param primitives list of primitives
+         */
+        Shadow(const std::vector<std::shared_ptr<Primitive::IPrimitive>>& primitives);
+
+        /**
+         * @brief Destroy the Shadow object
```
Comment (AxelF44) [inline include/Primitives/Shadow.hpp:31]:
```suggestion
         * @brief Destroy the Shadow object.
```

File: include/Primitives/Shadow.hpp [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+         *
+         */
+        ~Shadow() = default;
+
+        /**
+         * @brief return if there is a shadow
```
Comment (AxelF44) [inline include/Primitives/Shadow.hpp:37]:
```suggestion
         * @brief Return if there is a shadow.
```

File: include/Primitives/Sphere.hpp [hunk 10]
```diff
@@ -90,7 +81,15 @@ namespace Primitive {
              * @brief Get the Material object
```
Comment (AxelF44) [inline include/Primitives/Sphere.hpp:81]:
```suggestion
             * @brief Get the Material object.
```

File: include/Primitives/Sphere.hpp [hunk 11]
```diff
@@ -47,15 +47,6 @@ namespace Primitive {
...
-             * @brief compute the hit point color of a sphere
-             *
-             * @param hitPoint to compute the color
-             * @param lights container of lights
-             * @return Math::Point3D color
-             */
-            Math::Point3D computeColor(const Math::Point3D& hitPoint, const Light::LightsContainer& lights) const override;
-
             /**
              * @brief Set the Origin object
```
Comment (AxelF44) [inline include/Primitives/Sphere.hpp:51]:
```suggestion
             * @brief Set the Origin object.
```

File: include/Primitives/Sphere.hpp [hunk 12]
```diff
@@ -90,7 +81,15 @@ namespace Primitive {
              * @brief Get the Material object
              * @return Material of sphere
              */
-            std::shared_ptr<Material::IMaterial> getMaterial() const;
+            std::shared_ptr<Material::IMaterial> getMaterial() const override;
+
+            /**
+             * @brief Get the Normal of the object
```
Comment (AxelF44) [inline include/Primitives/Sphere.hpp:87]:
```suggestion
             * @brief Get the Normal of the object.
```

File: src/Core/Lights/LightsContainer.cpp [hunk 13]
```diff
@@ -22,12 +22,12 @@ std::vector<std::shared_ptr<Light::ILight>> Light::LightsContainer::getLightsLis
...
+Math::Point3D Light::LightsContainer::computeColor(Math::Vector3D primitiveNormal, const Math::Point3D& hitPoint, Math::Point3D color, const Primitives::Shadow& shadow) const
 {
     Math::Point3D highColor;
 
     for (size_t i = 0; i < _Lights.size(); i++) {
-        Math::Point3D colorFind = _Lights[i]->computeColor(primitiveNormal, hitPoint, color);
+        Math::Point3D colorFind = _Lights[i]->computeColor(primitiveNormal, hitPoint, color, shadow);
         if (highColor.length() < colorFind.length() ) {
             highColor = colorFind;
         }
```
Comment (AxelF44) [inline src/Core/Lights/LightsContainer.cpp:33]:
```suggestion
        if (highColor.length() < colorFind.length() )
            highColor = colorFind;
```

File: src/Core/Primitives/PrimitivesContainer.cpp [hunk 14]
```diff
@@ -7,6 +7,8 @@
 
 #include "Primitives/PrimitivesContainer.hpp"
 #include "RaytracerRules.hpp"
+#include "Materials/FlatColor.hpp"
+#include "Primitives/Shadow.hpp"
```
Comment (AxelF44) [inline src/Core/Primitives/PrimitivesContainer.cpp:11]:
```suggestion
#include "RaytracerRules.hpp"
#include "Materials/FlatColor.hpp"
#include "Primitives/Shadow.hpp"
#include "Primitives/PrimitivesContainer.hpp"
```

File: src/Core/Primitives/Shadow.cpp [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+** EPITECH PROJECT, 2024
+** Raytracer
+** File description:
+** Shadow
+*/
+
+#include "Primitives/Shadow.hpp"
+
+#include <iostream>
+#include <math.h>
```
Comment (AxelF44) [inline src/Core/Primitives/Shadow.cpp:11]:
```suggestion
#include <math.h>
#include <iostream>
```

File: src/Plugins/Lights/Ambient/Ambient.cpp [hunk 16]
```diff
@@ -44,8 +44,9 @@ Light::LightType Light::Ambient::getType(void) const
     return Light::LightType::ambient;
 }
 
-Math::Point3D Light::Ambient::computeColor(Math::Vector3D primitiveNormal, const Math::Point3D& hitPoint, Math::Point3D color) const
+Math::Point3D Light::Ambient::computeColor(Math::Vector3D primitiveNormal, const Math::Point3D& hitPoint, Math::Point3D color, const Primitives::Shadow& shadow) const
 {
+    (void) shadow;
     (void) primitiveNormal;
     (void) hitPoint;
```
Comment (AxelF44) [inline src/Plugins/Lights/Ambient/Ambient.cpp:51]:
```suggestion
    (void) shadow;
    (void) hitPoint;
    (void) primitiveNormal;
```

File: src/Plugins/Primitives/Sphere/Sphere.cpp [hunk 17]
```diff
@@ -80,3 +67,8 @@ std::shared_ptr<Material::IMaterial> Primitive::Sphere::getMaterial() const
 {
     return this->_material;
 }
+
+Math::Vector3D Primitive::Sphere::getNormal(const Math::Vector3D& hitPoint) const
+{
+    return Math::Vector3D (hitPoint.x() - _origin.x(), hitPoint.y() - _origin.y(), hitPoint.z() - _origin.z());
+}
```
Comment (AxelF44) [inline src/Plugins/Primitives/Sphere/Sphere.cpp:74]:
```suggestion
}

```

**Review by AxelF44 — CHANGES_REQUESTED** [review]

File: include/Lights/Ambient.hpp [hunk 18]
```diff
@@ -81,7 +81,7 @@ class Light::Ambient : public Light::ILight {
          * @param color color
          * @return Math::Point3D color
          */
-        Math::Point3D computeColor(Math::Vector3D primitiveNormal, const Math::Point3D& hitPoint, Math::Point3D color) const override;
+        Math::Point3D computeColor(Math::Vector3D primitiveNormal, const Math::Point3D& hitPoint, Math::Point3D color, const Primitives::Shadow& shadow) const override;
```
Comment (Njord201) [inline include/Lights/Ambient.hpp:84]: Describes the parameter added in the documentation.

File: include/Lights/Directional.hpp [hunk 19]
```diff
@@ -97,7 +97,7 @@ class Light::Directional : public Light::ILight {
          * @param color color
          * @return Math::Point3D color
          */
-        Math::Point3D computeColor(Math::Vector3D primitiveNormal, const Math::Point3D& hitPoint, Math::Point3D color) const override;
+        Math::Point3D computeColor(Math::Vector3D primitiveNormal, const Math::Point3D& hitPoint, Math::Point3D color, const Primitives::Shadow& shadow) const override;
```
Comment (Njord201) [inline include/Lights/Directional.hpp:100]: Describes the parameter added in the documentation.

File: include/Lights/ILight.hpp [hunk 20]
```diff
@@ -43,5 +44,5 @@ class Light::ILight {
              * @param color color
              * @return Math::Point3D color
              */
-            virtual Math::Point3D computeColor(Math::Vector3D primitiveNormal, const Math::Point3D& hitPoint, Math::Point3D color) const = 0;
+            virtual Math::Point3D computeColor(Math::Vector3D primitiveNormal, const Math::Point3D& hitPoint, Math::Point3D color, const Primitives::Shadow& shadow) const = 0;
```
Comment (Njord201) [inline include/Lights/ILight.hpp:47]: Describes the parameter added in the documentation.

File: include/Lights/LightsContainer.hpp [hunk 21]
```diff
@@ -59,7 +59,7 @@ namespace Light {
              * @param color color
              * @return Math::Point3D color
              */
-            Math::Point3D computeColor(Math::Vector3D primitiveNormal, const Math::Point3D& hitPoint, Math::Point3D color) const;
+            Math::Point3D computeColor(Math::Vector3D primitiveNormal, const Math::Point3D& hitPoint, Math::Point3D color, const Primitives::Shadow& shadow) const;
```
Comment (Njord201) [inline include/Lights/LightsContainer.hpp:62]: Describes the parameter added in the documentation.

File: include/Lights/Point.hpp [hunk 22]
```diff
@@ -82,7 +82,7 @@ class Light::Point : public Light::ILight {
          * @param color color
          * @return Math::Point3D color
          */
-        Math::Point3D computeColor(Math::Vector3D primitiveNormal, const Math::Point3D& hitPoint, Math::Point3D color) const override;
+        Math::Point3D computeColor(Math::Vector3D primitiveNormal, const Math::Point3D& hitPoint, Math::Point3D color, const Primitives::Shadow& shadow) const override;
```
Comment (Njord201) [inline include/Lights/Point.hpp:85]: Describes the parameter added in the documentation.

File: include/Primitives/IPrimitive.hpp [hunk 23]
```diff
@@ -40,12 +42,18 @@ namespace Primitive {
...
+             *
+             * @param hitPoint to have the normal
```
Comment (Njord201) [inline include/Primitives/IPrimitive.hpp:47]:
```suggestion
             * @param hitPoint to compute the normal
```

File: include/Primitives/IPrimitive.hpp [hunk 24]
```diff
@@ -40,12 +42,18 @@ namespace Primitive {
...
+             * @return Math::Vector3D
+             */
+            virtual Math::Vector3D getNormal(const Math::Vector3D& hitPoint) const = 0;
+
+            /**
+             * @brief Get the Material object
```
Comment (Njord201) [inline include/Primitives/IPrimitive.hpp:53]:
```suggestion
             * @brief Get the Material object.
```

File: include/Primitives/Sphere.hpp [hunk 25]
```diff
@@ -90,7 +81,15 @@ namespace Primitive {
...
+             *
+             * @param hitPoint to compute
```
Comment (Njord201) [inline include/Primitives/Sphere.hpp:89]:
```suggestion
             * @param hitPoint to compute the normal
```

File: src/Core/Primitives/PrimitivesContainer.cpp [hunk 26]
```diff
@@ -43,8 +45,20 @@ Math::Point3D Primitive::PrimitivesContainer::hitPoint(const Raytracer::Ray& ray
...
+
+Math::Point3D Primitive::PrimitivesContainer::computeColor(const std::shared_ptr<Primitive::IPrimitive>& primitive, const Math::Point3D& hitPoint, const Light::LightsContainer& lights) const
+{
+    Math::Vector3D normal = primitive->getNormal(hitPoint);
+
+    if (primitive->getMaterial()->getType() == Material::MaterialType::FlatColor) {
+        std::shared_ptr<FlatColor> primitiveFlatColor = std::dynamic_pointer_cast<FlatColor>(primitive->getMaterial());
+        return lights.computeColor(normal, hitPoint, Math::Point3D(primitiveFlatColor->getR(), primitiveFlatColor->getG(), primitiveFlatColor->getB()), Primitives::Shadow(_primitives));
+    }
+    std::cout << "material not handle in sphere" << std::endl;
```
Comment (Njord201) [inline src/Core/Primitives/PrimitivesContainer.cpp:60]:
```suggestion
    std::cout << "Material not handle in sphere" << std::endl;
```

File: src/Core/Primitives/Shadow.cpp [hunk 27]
```diff
@@ -0,0 +1,45 @@
...
+
+Primitives::Shadow::Shadow(const std::vector<std::shared_ptr<Primitive::IPrimitive>>& primitives) : _primitives(primitives) {};
+
+static double myRound(double v, double precision)
```
Comment (Njord201) [inline src/Core/Primitives/Shadow.cpp:15]: What is 'v' ?

File: src/Core/Primitives/Shadow.cpp [hunk 28]
```diff
@@ -0,0 +1,45 @@
...
+{
+    return std::round(v/precision)*precision;
+}
+
+bool Primitives::Shadow::isShadow(const Math::Vector3D& vectorLightToPoint, const Math::Point3D& hitPoint) const
+{
+    Raytracer::Ray ray(hitPoint, vectorLightToPoint);
+    Math::Point3D h = hitPoint;
```
Comment (Njord201) [inline src/Core/Primitives/Shadow.cpp:23]: What means h ?

File: src/Core/Primitives/Shadow.cpp [hunk 29]
```diff
@@ -0,0 +1,45 @@
...
+#include <math.h>
+
+Primitives::Shadow::Shadow(const std::vector<std::shared_ptr<Primitive::IPrimitive>>& primitives) : _primitives(primitives) {};
+
+static double myRound(double v, double precision)
+{
+    return std::round(v/precision)*precision;
+}
+
+bool Primitives::Shadow::isShadow(const Math::Vector3D& vectorLightToPoint, const Math::Point3D& hitPoint) const
```
Comment (Njord201) [inline src/Core/Primitives/Shadow.cpp:20]: I don't think there's any need to create these variables (ray, h, vltp)

File: src/Plugins/Lights/Ambient/Ambient.cpp [hunk 30]
```diff
@@ -44,8 +44,9 @@ Light::LightType Light::Ambient::getType(void) const
     return Light::LightType::ambient;
 }
 
-Math::Point3D Light::Ambient::computeColor(Math::Vector3D primitiveNormal, const Math::Point3D& hitPoint, Math::Point3D color) const
+Math::Point3D Light::Ambient::computeColor(Math::Vector3D primitiveNormal, const Math::Point3D& hitPoint, Math::Point3D color, const Primitives::Shadow& shadow) const
 {
+    (void) shadow;
```
Comment (Njord201) [inline src/Plugins/Lights/Ambient/Ambient.cpp:49]: Add TODO comment and create an Issue (refer the issue in the comment)

File: src/Plugins/Lights/Directional/Directional.cpp [hunk 31]
```diff
@@ -59,13 +59,18 @@ Light::LightType Light::Directional::getType(void) const
...
-Math::Point3D Light::Directional::computeColor(Math::Vector3D primitiveNormal, const Math::Point3D& hitPoint, Math::Point3D color) const
+Math::Point3D Light::Directional::computeColor(Math::Vector3D primitiveNormal, const Math::Point3D& hitPoint, Math::Point3D color, const Primitives::Shadow& shadow) const
 {
-    (void) hitPoint;
-    Math::Vector3D pr = primitiveNormal / primitiveNormal.length();
+    Math::Vector3D normal = primitiveNormal / primitiveNormal.length();
     Math::Vector3D dir = getDirection() / getDirection().length();
 
-    color *= -pr.dot(dir);
+    // TODO : rewrite method to handle the position of the
```
Comment (Njord201) [inline src/Plugins/Lights/Directional/Directional.cpp:67]: Open an issue

File: src/Plugins/Primitives/Cylinder/Cylinder.cpp [hunk 32]
```diff
@@ -110,3 +98,9 @@ void Primitive::Cylinder::setMaterial(std::shared_ptr<Material::IMaterial> mater
 {
     this->_material = material;
 }
+
+Math::Vector3D Primitive::Cylinder::getNormal(const Math::Vector3D& hitPoint) const
+{
+    //TODO : fix the computation of the cylinder normal
```
Comment (Njord201) [inline src/Plugins/Primitives/Cylinder/Cylinder.cpp:104]: Open an issue

**Review by Njord201 — CHANGES_REQUESTED** [review]

File: include/Lights/Directional.hpp [hunk 33]
```diff
@@ -95,9 +95,10 @@ class Light::Directional : public Light::ILight {
          * @param primitiveNormal normal to the hitpoint
```
Comment (AxelF44) [inline include/Lights/Directional.hpp:95]:
```suggestion
         * @param primitive Normal normal to the hitpoint
```

File: include/Lights/LightsContainer.hpp [hunk 34]
```diff
@@ -20,46 +20,47 @@ namespace Light {
...
              *
              * @return std::vector<std::shared_ptr<Light::ILight>>
              */
             std::vector<std::shared_ptr<Light::ILight>> getLightsList(void) const;
 
             /**
-             * @brief compute the color point with lights
+             * @brief Compute the color point with lights.
              *
              * @param primitiveNormal normal to the hitpoint
```
Comment (AxelF44) [inline include/Lights/LightsContainer.hpp:57]:
```suggestion
             * @param primitive Normal normal to the hitpoint
```

File: include/Primitives/Plane.hpp [hunk 35]
```diff
@@ -25,75 +25,74 @@ namespace Primitive {
...
             /**
-             * @brief Set the Material
+             * @brief Set the Material.
              * @param material New material to set
              */
             void setMaterial(std::shared_ptr<Material::IMaterial> material);
 
             /**
-             * @brief compute the hit point color of a sphere
+             * @brief Get the Normal of the object.
```
Comment (AxelF44) [inline include/Primitives/Plane.hpp:90]:
```suggestion
             * @brief Get the normal of the object.
```

**Review by theophile-jr — APPROVED** [review]

**Review by AxelF44 — CHANGES_REQUESTED** [review]

**Review by AxelF44 — APPROVED** [review]

**Review by Njord201 — DISMISSED** [review]

**Comment by AxelF44** [comment]
Hello Mathieu,

Can you resolve your conflict please !

**Comment by Njord201** [comment]
It's becoming urgent

> Hello Mathieu,
> 
> Can you resolve your conflict please !

**Review by AxelF44 — APPROVED** [review]

File: include/Primitives/Plane.hpp [hunk 36]
```diff
@@ -25,75 +25,74 @@ namespace Primitive {
...
+             * @brief Set the Position object.
              *
              * @param position Position to set
              */
             void setPosition(Math::Point3D position);
 
             /**
-             * @brief Get the Material object
+             * @brief Get the Material object.
              * @return Material of plane
```
Comment (Njord201) [inline include/Primitives/Plane.hpp:79]:
```suggestion

             * @return Material of plane
```

File: include/Primitives/Plane.hpp [hunk 37]
```diff
@@ -25,75 +25,74 @@ namespace Primitive {
...
              */
-            std::shared_ptr<Material::IMaterial> getMaterial() const;
+            std::shared_ptr<Material::IMaterial> getMaterial() const override;
 
             /**
-             * @brief Set the Material
+             * @brief Set the Material.
              * @param material New material to set
```
Comment (Njord201) [inline include/Primitives/Plane.hpp:85]:
```suggestion

             * @param material New material to set
```

File: include/Primitives/Sphere.hpp [hunk 38]
```diff
@@ -34,63 +34,63 @@ namespace Primitive {
...
              * @brief Set the Origin object
+             *
              * @param origin New origin to set
              */
             void setOrigin(Math::Point3D origin);
 
             /**
-             * @brief Set the Radius
+             * @brief Set the Radius.
              * @param radius New radius to set
```
Comment (Njord201) [inline include/Primitives/Sphere.hpp:59]:
```suggestion

             * @param radius New radius to set
```

File: include/Primitives/Sphere.hpp [hunk 39]
```diff
@@ -34,63 +34,63 @@ namespace Primitive {
...
              */
             void setRadius(double radius);
 
             /**
-             * @brief Set the Material
+             * @brief Set the Material.
              * @param material New material to set
```
Comment (Njord201) [inline include/Primitives/Sphere.hpp:65]:
```suggestion

             * @param material New material to set
```

File: include/Primitives/Sphere.hpp [hunk 40]
```diff
@@ -34,63 +34,63 @@ namespace Primitive {
...
              */
             void setMaterial(std::shared_ptr<Material::IMaterial> material);
 
             /**
-             * @brief Get the Origin object
+             * @brief Get the Origin object.
              * @return Origin of sphere
```
Comment (Njord201) [inline include/Primitives/Sphere.hpp:71]:
```suggestion

             * @return Origin of sphere
```

File: include/Primitives/Sphere.hpp [hunk 41]
```diff
@@ -34,63 +34,63 @@ namespace Primitive {
...
              */
             Math::Point3D getOrigin() const;
 
             /**
-             * @brief Get the Origin object
+             * @brief Get the Origin object.
              * @return Radius of sphere
```
Comment (Njord201) [inline include/Primitives/Sphere.hpp:77]:
```suggestion

             * @return Radius of sphere
```

File: include/Primitives/Sphere.hpp [hunk 42]
```diff
@@ -34,63 +34,63 @@ namespace Primitive {
...
              */
             double getRadius() const;
 
             /**
-             * @brief Get the Material object
+             * @brief Get the Material object.
              * @return Material of sphere
```
Comment (Njord201) [inline include/Primitives/Sphere.hpp:83]:
```suggestion

             * @return Material of sphere
```

**Review by Njord201 — CHANGES_REQUESTED** [review]

File: include/Lights/LightsContainer.hpp [hunk 43]
```diff
@@ -21,46 +21,49 @@ namespace Light {
...
              * @param hitPoint hitpoint
              * @param color color
+             * @param shadow Primitive::Shadow class to handle shadows
+             * @return Math::Point3D color
+             *
              * @return Color color
              */
-            Color computeColor(Math::Vector3D primitiveNormal, const Math::Point3D& hitPoint, Math::Point3D color) const;
+            Color computeColor(Math::Vector3D primitiveNormal, const Math::Point3D& hitPoint, Math::Point3D color, const Primitives::Shadow& shadow) const;
```
Comment (Njord201) [inline include/Lights/LightsContainer.hpp:66]:
```suggestion
            Color computeColor(Math::Vector3D primitiveNormal, const Math::Point3D& hitPoint, Math::Point3D color, const Primitives::Shadow& shadow) const override;
```

File: include/Primitives/Cone.hpp [hunk 44]
```diff
@@ -113,6 +104,14 @@ namespace Primitive {
              */
             std::shared_ptr<Material::IMaterial> getMaterial() const;
 
+            /**
+             * @brief Get the Normal of the object.
+             *
+             * @param hitPoint to have the normal
+             * @return Math::Vector3D
+             */
+            Math::Vector3D getNormal(const Math::Vector3D& hitPoint) const;
```
Comment (Njord201) [inline include/Primitives/Cone.hpp:113]:
```suggestion
            Math::Vector3D getNormal(const Math::Vector3D& hitPoint) const override;
```

File: include/Primitives/Cylinder.hpp [hunk 45]
```diff
@@ -81,18 +72,27 @@ namespace Primitive {
...
              */
             void setMaterial(std::shared_ptr<Material::IMaterial> material);
+
+            /**
+             * @brief Get the Normal of the object.
+             *
+             * @param hitPoint to have the normal
+             * @return Math::Vector3D
+             */
+            Math::Vector3D getNormal(const Math::Vector3D& hitPoint) const;
```
Comment (Njord201) [inline include/Primitives/Cylinder.hpp:94]:
```suggestion
            Math::Vector3D getNormal(const Math::Vector3D& hitPoint) const override;
```

File: include/Primitives/Plane.hpp [hunk 46]
```diff
@@ -25,75 +25,76 @@ namespace Primitive {
...
+             * @brief Get the normal of the object.
              *
-             * @param hitPoint to compute the color
-             * @param lights container of lights
-             * @return Color color
+             * @param hitPoint to have the normal
+             * @return Math::Vector3D
              */
-            Color computeColor(const Math::Point3D& hitPoint, const Light::LightsContainer& lights) const override;
+            Math::Vector3D getNormal(const Math::Vector3D& hitPoint) const;
```
Comment (Njord201) [inline include/Primitives/Plane.hpp:97]:
```suggestion
            Math::Vector3D getNormal(const Math::Vector3D& hitPoint) const override;
```

File: include/Primitives/PrimitivesContainer.hpp [hunk 47]
```diff
@@ -53,12 +54,22 @@ namespace Primitive {
...
 
+            /**
+             * @brief Compute the color pixel of a primitive's hitpoint.
+             *
+             * @param primitive primitive to compute
+             * @param hitPoint to check
+             * @param lights list of lights
+             * @return Math::Point3D
+             */
+            Color computeColor(const std::shared_ptr<Primitive::IPrimitive>& primitive, const Math::Point3D& hitPoint, const Light::LightsContainer& lights) const;
```
Comment (Njord201) [inline include/Primitives/PrimitivesContainer.hpp:71]:
```suggestion
            Color computeColor(const std::shared_ptr<Primitive::IPrimitive>& primitive, const Math::Point3D& hitPoint, const Light::LightsContainer& lights) const override;
```

File: include/Primitives/Sphere.hpp [hunk 48]
```diff
@@ -34,63 +34,68 @@ namespace Primitive {
...
-            std::shared_ptr<Material::IMaterial> getMaterial() const;
+            std::shared_ptr<Material::IMaterial> getMaterial() const override;
+
+            /**
+             * @brief Get the Normal of the object.
+             *
+             * @param hitPoint to compute the normal
+             * @return Math::Vector3D
+             */
+            Math::Vector3D getNormal(const Math::Vector3D& hitPoint) const;
```
Comment (Njord201) [inline include/Primitives/Sphere.hpp:98]:
```suggestion
            Math::Vector3D getNormal(const Math::Vector3D& hitPoint) const override;
```

**Review by Njord201 — CHANGES_REQUESTED** [review]
Override...

File: include/Lights/LightsContainer.hpp (hunk 43 above)
Comment (mathieurobert1) [reply include/Lights/LightsContainer.hpp:66]: I do not agree, LightContainer is not a children class of ILight

**Review by theophile-jr — APPROVED** [review]

**Review by AxelF44 — APPROVED** [review]

**Pull request merged**